### ISSUE TITLE
Guide for manually rotating control plane needs more information

### ISSUE THREAD

**Issue by christianhuening** [linkerd/linkerd2#4287]
### What is the issue?
I think, the docs for rotating the TLS Credentials manually (https://linkerd.io/2/tasks/manually-rotating-control-plane-tls-credentials/) do lack an important pointer towards the necessity of rotating all proxy side-cars OR wait 24h before removing the old ca cert from the bundle? 

### What happened?
We rotated the Trust Anchor, kicked over all our side-car proxies. Then we issued a new Identity Issuer Cert and applied that. All was fine.
Then we followed on with the guide and removed the old CA cert from the bundle. That caused Proxies to fail with BadCertificate errors. We re-issued the CA Bundle and all was fine again. 

This caused about a 12 minute partial-downtime.

### Possible solution
So my takeaway is, that BEFORE removing the old CA you EITHER have to rotate all proxies once more OR wait 24h until the Identity Issuer has renewed all Side-Car identies.
Is that assumption right?  If so, we should update the docs.

Linkerd Version: 2.7.1
Kubernetes: 1.15.11

**Comment by grampelberg** [comment]
@zaharidichev WDYT?

**Comment by zaharidichev** [comment]
I tried replicating the problem and could not see the same behavior but I have of idea of how to explore that a bit more. Will follow up with a solution soon

**Comment by christianhuening** [comment]
ping

**Comment by ihcsim** [comment]
@ThomWright Thanks for the detailed steps. The bundle that contains the new and old trusts should ensure that old and new proxies can communicate with each other after step 5. After deploying the new issuer cert in step 4, did the identity service issue an `IssuerUpdated` event? E.g.,

```sh
kubectl get events --field-selector reason=IssuerUpdated -n linkerd

LAST SEEN   TYPE     REASON          OBJECT                        MESSAGE
9s          Normal   IssuerUpdated   deployment/linkerd-identity   Updated identity issuer
```

Just wanna see if there is a pattern to look out for in our repro.

**Comment by ThomWright** [comment]
Pretty sure the `IssuerUpdated` event was emitted, yes.

>The bundle that contains the new and old trusts should ensure that old and new proxies can communicate with each other after step 5

But that bundle no longer exists, except in the old proxies. From my understanding, the old proxies will trust certs signed by either issuer. The new proxies will not, right?. What am I missing here?

Another thing to point out: after step 5, I assume the control plane will no longer trust the certificates in the non-control-plane proxies. The control plane only has the new trust anchor, while all other proxies are using certificates signed by another issuer. I assume that's why we immediately started seeing `BadCertificate` errors in our control plane after step 5.

It seems to me that the correct steps are:

- upgrade the control plane to use the new issuer certificate
- restart all proxies, so they have certificates signed by the new issuer
- upgrade the control plane to use the new trust anchor
- restart all proxies at leisure

**Comment by ihcsim** [comment]
@christianhuening @ThomWright Sorry for the delayed response. I can confirm that this problem is reproducible. 

TL;DR as you folks pointed out, the proxies in the data plane need to be restarted after the new issuer certificate is deployed, and before replacing the bundled trust with only the new trust anchor. Otherwise, the communication between the existing proxies and control plane will fail. I did notice that the traffic between all existing proxies still work though. I will update the doc accordingly.

My repro findings: Following the instructions in the doc, if I replace the bundled trust with the new trust anchor, with the proxies still using certs issued by an old issuer, I can see these errors in the log.

`linkerd check --proxy` will hang:

```sh
$ linkerd check --proxy
# ...
linkerd-identity-data-plane
---------------------------
‼ data plane proxies certificate match CA
    Some pods do not have the current trust bundle and must be restarted: # notice these proxies haven't been restarted
        * emojivoto/emoji-78668f9dff-htg2w
        * emojivoto/vote-bot-5d87f59c56-kl4m2
        * emojivoto/voting-7bcc68d4b4-t99h5
        * emojivoto/web-7578cd6b8d-6d67z
    see https://linkerd.io/checks/#l5d-identity-data-plane-proxies-certs-match-ca for hints
# ....
linkerd-data-plane
------------------
√ data plane namespace exists
√ data plane proxies are ready
/ waiting for check to complete   # this hangs due to errors in the proxies
```

The Prometheus proxy shows error lots of errors when attempting to scrape the emojivoto app:

```sh
$ k -n linkerd logs linkerd-prometheus-569cbc7ccf-dvrcg linkerd-proxy 
[   203.783253157s]  WARN endpoint{peer.addr=10.244.0.28:4191}: rustls::session: Sending fatal alert BadCertificate
[   203.945536404s]  WARN endpoint{peer.addr=10.244.0.27:4191}: rustls::session: Sending fatal alert BadCertificate
[   203.993604409s]  WARN endpoint{peer.addr=10.244.0.29:4191}: rustls::session: Sending fatal alert BadCertificate
[   204.48592638s]  WARN endpoint{peer.addr=10.244.0.30:4191}: rustls::session: Sending fatal alert BadCertificate
[   204.285610733s]  WARN endpoint{peer.addr=10.244.0.28:4191}: rustls::session: Sending fatal alert BadCertificate
```

The voting service also logged many `BadCertificate` errors related to requests from Prometheus:

```sh
k -n emojivoto logs voting-7bcc68d4b4-t99h5 linkerd-proxy
# ...
[   867.279650960s] ERROR daemon:admin{listen.addr=0.0.0.0:4191}:accept{peer.addr=10.244.0.39:55598}: rustls::session: TLS alert received: Message {
    typ: Alert,
    version: TLSv1_2,
    payload: Alert(
        AlertMessagePayload {
            level: Fatal,
            description: BadCertificate,
        },
    ),
} 
# ...
```

This problem doesn't happen if the data plane is restarted after the control plane is upgraded with a new issuer, but before rotating out the bundled trust.
